twister: Add options deciding if paths be included in tests' names

PerMac opened the pull request:
Test suites names are not being handled uniformly for tests not in zephyr tree. Their names depend on -T arg used in twister's CLI. The newly added options allow to select if twister should add paths to suite names. This is needed if test plans are to be used for tests outside of zephyr tree.

PerMac committed:
twister: Add options deciding if paths be included in tests' names

Test suites names are not being handled uniformly for tests not
in zephyr tree. Their names depend on -T arg used in twister's CLI.
The newly added options allow to select if twister should add paths
to suite names. This is needed if test plans are to be used for tests
outside of zephyr tree.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac commented:
As requested, separated from #61838

hakehuang commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -437,6 +437,10 @@ def add_parse_arguments(parser = None):
              "faster compilation since builds will be incremental.")
 
+    parser.add_argument(
+        "--no-path-name", action="store_true",
```
may be we can put all those into a concise report mode, and combinet several opetion into one

PerMac replied:
sorry, I don't follow. Which "all those"? I am adding a single arg which changes the way the suite names are created (not adding path). I don't think this option can be combined in one with others.

PerMac replied:
what do you mean by "option"? A grouping for args? How would it make the interface simpler? What is an "internal control flag"? The arg I am adding can be used with any other argument. It not only change the internal handling of names but also affects the names in the output report, so to me it is not really an "internal" argument. What other arguments should be in this `concise-report`? Sorry, I really don't follow. If it is a general comment about idea on modification of twister interface, maybe you can open an RFC where show what you'd like to achieve?

hakehuang replied:
let's discuss this in weekly meeting. I mean maybe we can have some default report options:
--verbose_mode
--concise_mode

and in each mode  a flags dict is build in like
concise_mode = {
'short_path' => true,
'log_time' => false,
'log_in_line' => false,
}

and those flags in the dict are mapped  to twister internal control logic

nashif reviewed (changes requested):
- wondering if this should just be made the default, otherwise we will end up with different reports and test identifiers depending on whether you have used the option or now.
- The option name on its own provides no context, so maybe be a bit verbose in the option name
- There is a loss of information when this option is used, we no longer have the path anywhere in the generated reports:

```
        {
            "name":"kernel.threads.init",
            "arch":"arm64",
            "platform":"rcar_salvator_xs_m3",
            "run_id":"c43467c33c71b9d51376c6aec53784aa",
            "runnable":false,
            "retries":0,
            "status":"passed",
            "execution_time":"0.00",
            "testcases":[
                {
                    "identifier":"kernel.threads.init.kinit_coop_thread",
                    "execution_time":"0.00",
                    "status":null
                },
                {
                    "identifier":"kernel.threads.init.kinit_preempt_thread",
                    "execution_time":"0.00",
                    "status":null
                },
                {
                    "identifier":"kernel.threads.init.kdefine_coop_thread",
                    "execution_time":"0.00",
                    "status":null
                },
                {
                    "identifier":"kernel.threads.init.kdefine_preempt_thread",
                    "execution_time":"0.00",
                    "status":null
                }
            ]
        },
```

if something fails, how do we know which test application was it part of ?


- In the final text where list of failures is displayed, we get somethng like this:

```
    - The following issues were found (showing the top 10 items):
INFO    - 1) kernel.threads.init on qemu_cortex_a53 failed (Failed)
INFO    - 2) kernel.threads.init on qemu_cortex_a53_smp failed (Failed)
INFO    - 3) kernel.threads.init on qemu_riscv64_smp failed (Failed)
INFO    - 4) kernel.threads.init on qemu_riscv64 failed (Failed)
INFO    - 5) kernel.threads.init on qemu_riscv32 failed (Failed)
INFO    - 6) kernel.threads.init on qemu_riscv32_smp failed (Failed)
INFO    - 7) kernel.threads.init on qemu_riscv32e failed (Failed)
INFO    - 8) kernel.threads.init on m2gl025_miv failed (Failed)
INFO    - 9) kernel.threads.init on qemu_xtensa failed (Failed)
INFO    - 10) kernel.threads.init on qemu_nios2 failed (Failed)
INFO    -
INFO    - To rerun the tests, call twister using the following commandline:
INFO    - west twister -p <PLATFORM> -s <TEST ID>, for example:
INFO    -
INFO    - west twister -p qemu_nios2 -s kernel.threads.init
INFO    - or with west:
INFO    - west build -p -b qemu_nios2 -T kernel.threads.init
INFO    - -+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

`west build -p -b qemu_nios2 -T kernel.threads.init` and `west twister -p qemu_nios2 -s kernel.threads.init` will not work

PerMac commented:
I am fine with having this as default. We already were discussing such change but I don't know if a consensus was reached. Won't it affect how data is parsed and added to the elastic search?  Also it will affect some other options as well, like the usage of -s will have to change. There is a real issue: twister is not properly handling names of tests outside of zephyr tree. Last time I tried to fix this by modifying the way the names are created it backlashed and was revered, hence now I propose such change as an option.

I will add more verbose description.

I will add a commit that adds a path to the report

PerMac committed:
twister: Add "path" entry to json test report

"Path" is a path (relative to zephyr) to a directory with test suite's
definition. Such entry helps to locate a given test.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac commented:
@nashif I am done with changes to this PR for now (of course I will address new comments). Please take a look.

- should it be default:  I also opened an RFC with proposal of such transition https://github.com/zephyrproject-rtos/zephyr/issues/62926
- more verbose descriptions added
- now includes a commit which adds 'path' entry to reports
- synopsis modified: recommendation for west build will always be the same, where source dir is explicitly given and -T points to a name without path; twister command will have `--no-path-name` added at the end if this option was used

nashif commented on scripts/pylib/twister/twisterlib/reports.py:
```diff
@@ -434,11 +436,16 @@ def synopsis(self):
             logger.info("")
             logger.info("To rerun the tests, call twister using the following commandline:")
-            logger.info("west twister -p <PLATFORM> -s <TEST ID>, for example:")
+            cmd_west_twist_examp = f"west twister -p <PLATFORM> -s <TEST ID>"
```
I am not a big fine of such abbriviations, i.e. twist for twister or examp for example, please either make it complete or figure out a shorter variable name, IMO this could be much shorter and since this is a local variable

PerMac replied:
I changed to `cmd_example` and `cmd_explicite`. Are these names ok?

PerMac commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -437,6 +437,21 @@ def add_parse_arguments(parser = None):
              "faster compilation since builds will be incremental.")
 
+    parser.add_argument(
+        '--path-name', action='store_true',
```
Sorry, my bad. I read it wrongly and thought your comment was about description in help. I changed the flag to `--detailed-test-id` and `--no-detailed-test-id`. Are these names ok?

gchwier reviewed (commented):
please fix compliance checks also

gchwier commented on scripts/pylib/twister/twisterlib/reports.py:
```diff
@@ -434,11 +436,16 @@ def synopsis(self):
             logger.info(f"or with west:")
-            logger.info(f"west build -p -b {example_instance.platform.name} -T {example_instance.testsuite.name}")
+            logger.info(f"west build -p -b {example_instance.platform.name} {example_instance.testsuite.source_dir_rel} -T {example_instance.testsuite.id}")
             logger.info("-+" * 40)
```
```suggestion
        if cnt and example_instance:
            logger.info("")
            logger.info("To rerun the tests, call twister using the following commandline:")
            extra_parameters = '' if detailed_test_id else ' --no-detailed-test-id'
            logger.info(f"west twister -p <PLATFORM> -s <TEST ID>{extra_parameters}, for example:")
            logger.info("")
            logger.info(f"west twister -p {example_instance.platform.name} -s {example_instance.testsuite.name}"
                        f"{extra_parameters}")
            logger.info(f"or with west:")
            logger.info(f"west build -p -b {example_instance.platform.name} "
                        f"{example_instance.testsuite.source_dir_rel} -T {example_instance.testsuite.name}")
            logger.info("-+" * 40)
```
will receive the same log, but with less changes in the code

gchwier commented on scripts/pylib/twister/twisterlib/testsuite.py:
```diff
@@ -391,10 +391,14 @@ def __init__(self, suite_root, suite_path, name, data=None):
         workdir = os.path.relpath(suite_path, suite_root)
-        self.name = self.get_unique(suite_root, workdir, name)
+
+        assert self.check_suite_name(name, suite_root, workdir)
```
Assert should only be used for testing and debugging — not in production environments.
Here the assert can be removed, because in `check_suite_name` an exception is raised.

gchwier commented on scripts/pylib/twister/twisterlib/testsuite.py:
```diff
@@ -447,10 +451,14 @@ def get_unique(testsuite_root, workdir, name):
                     """
                     )
-        return unique
+        return True
```
return can be removed. In case of failed check, `TwisterException` is raised

nashif reviewed (approved):
need to fix the issue raised by @gchwier

gchwier reviewed (approved):
`assert` mentioned before is not a big problem and this can be changed in a separate PR.
If the current PR can be merged, I agree
